src/Language/DifferentialDatalog/DatalogProgram.hs

```diff
@@ -307,3 +311,221 @@ progMirrorInputRelations d prefix =
+                                                                  exprArgs = [ E EVar { exprPos = nopos,
+                                                                                        exprVar = "g" } ] } },
+                        exprRight = E ETuple { exprPos = nopos,
+                                               exprTupleFields = [ E EVar { exprPos = nopos,
+                                                                            exprVar = "inputs" },
+                                                                   E EApply { exprPos = nopos,
+                                                                              exprFunc = origFname,
+                                                                              exprArgs = [ E EVar { exprPos = nopos,
+                                                                                                    exprVar = "original_group" } ] } ] } }
+    function = Function { funcPos = nopos,
+                          funcAttrs = [],
+                          funcName = fname,
+                          funcArgs = [ FuncArg { argPos = nopos,
+                                                 argName = "g",
+                                                 argMut = False,
+                                                 argType = TOpaque { typePos = nopos,
+                                                                     typeName = "std.Group",
+                                                                     typeArgs = [ TVar { typePos = nopos, tvarName = "K" },
+                                                                                  TTuple { typePos = nopos,
+                                                                                           typeTupArgs = [ TVar { typePos = nopos, tvarName = "I" },
+                                                                                                           TVar { typePos = nopos, tvarName = "V" } ] } ] } } ],
+                          funcType = TTuple { typePos = nopos,
+                                              typeTupArgs = [ TOpaque { typePos = nopos,
+                                                                        typeName = "std.Set",
+                                                                        typeArgs = [ TVar { typePos = nopos, tvarName = "I" } ] },
+                                                              TVar { typePos = nopos, tvarName = "V" } ] },
+                          funcDef = Just funcBody }
+  in M.insert fname function functions
+
+-- Generate and insert into the map of functions a wrapper aggregate function for
+-- each aggregate function used in the rule.
+updateFunctions :: [Rule] -> M.Map String Function -> M.Map String Function
+updateFunctions rules functions =
+  let
+    aggregates = filter rhsIsAggregate $ concatMap (\rule -> ruleRHS rule) rules
+  in case length aggregates of
+     0 -> functions
```

Comment:
I don't think you need a special case for 0; `foldl` will do the right thing if `aggregates` is empty.
Also, it is always recommended to use `foldl'` over `foldl` in Haskell.
Final nit: `concatMap (\rule -> ruleRHS rule) rules` is just `concatMap ruleRHS rules`.